### ISSUE TITLE
[patch #9816] Implement new programmer type: linuxspi

### ISSUE THREAD

**Issue by avrdude-admin** [avrdudes/avrdude#743]
Fri 14 Jun 2019 06:30:19 PM UTC

This adds a new linuxspi programmer type. Analogously to linuxgpio, this
programmer uses the Linux-specific spidev device.

For accessing the reset pin, this programmer uses an additional GPIO pin using
the linux/gpio.h API. In contrast to the stateless and deprecated sysfs
interface, linux/gpio.h uses proper file descriptors for every pin, and
performs housekeeping when the device is closed. That saves some lines of code.

The programmer can be used as follows:
avrdude -c linuxspi -P /dev/spidev:/dev/gpiochip[:resetpin] [...]

Separated by colons, the first part of the port (-P) argument denotes the
spidev device, the second one is the gpiochip that contains the reset pin, and
the optional third argument contains the pin number of the reset pin.

If not specified, the pin number defaults to the reset pin number defined in
avrdude.conf. Currently, this defaults to 25, which is connected to Pin 22 on a
Raspberry Pi.

Additionally, add some documentation.

Credits go to Kevin Cuzner, who started this fork ~5 years ago. I picked it up,
rebased it to the current mainline, massaged some parts of the code here and
there, fixed some minor bugs and switched to the linux/gpio.h GPIO accessors.

Signed-off-by: Ralf Ramsauer <ralf@vmexit.de>

[file #47099: avrdude-linuxspi-V1.patch](https://savannah.nongnu.org/patch/download.php?file_id=47099)

This issue was migrated from https://savannah.nongnu.org/patch/?9816

**Comment by avrdude-admin** [comment]
Fabian Viöl <sh3rm4n>
Tue 10 Mar 2020 03:30:25 PM UTC

Just wanted to leave a comment to say, that I am using this to flash an
atmega328p via SPI from the raspberry pi and it works quite well. Did not
have any issues with it, yet :)

**Comment by avrdude-admin** [comment]
Ralf Ramsauer <rralf>
Tue 10 Mar 2020 03:44:42 PM UTC

Fabian,

thank you very much for your comment and for testing my/our patch. I'd really wish that it would be integrated, but, to be honest, there was almost no reaction at all and avrdude seems to unfortunately be dead from a maintainer perspective. Sigh...

Ralf

**Comment by avrdude-admin** [comment]
Joerg Wunsch <joerg_wunsch>
Sat 19 Sep 2020 09:34:01 PM UTC

Applied in r1447, thanks! (Without own tests, don't have an RPi readily available where I could compile and use it.)

Thanks in particular for also providing documentation for it.

**Comment by avrdude-admin** [comment]
Lars Ollén <laeffe>
Tue 22 Sep 2020 08:20:39 PM UTC

I've used this patch for a while now, had some initial issues which I tracked down to this patch not setting the SPI mode, and if you are using SPI to communicate with as well, setting a incompatible SPI mode will cause this programmer being unable to communicate with the chip.

I fixed it by adding:
uint32_t mode = SPI_MODE_0 | SPI_NO_CS;
ret = ioctl(fd_spidev, SPI_IOC_WR_MODE32, &mode);
if (ret == -1) {
avrdude_message(MSG_INFO, "%s: error: Unable to set SPI mode %0X on %s\n", progname, mode, spidev);
goto close_spidev;
}

Before:
fd_gpiochip = open(gpiochip, 0);

in linuxspi.c

Ya I know that's the worst patch ever :( I applied this to the 6.3 source and added a shit load of debug messages to figure out where the issue was, so my local source is a mess ;)

**Comment by avrdude-admin** [comment]
Joerg Wunsch <joerg_wunsch>
Sat 27 Nov 2021 08:22:59 PM UTC

I also added Lars' patch in r1498.

I included a default port string now (the one which appears to be the default for my Raspberry Pi - yeah, finally able to test it), and I moved the ISP clock setting to the -B option as it is done for all other programmers. -b used to be used for RS232-attached programming dongles to really tell the RS232 baudrate. Additional advantage: -B can parse kHz and MHz suffix.

Finally, I changed the default bitclock to 200 kHz (so it fits the 1 MHz default CPU clock of most AVRs), improved documentation a bit (the texinfo file now has a full wiring example), and I implemented the configure logic in avrdude.conf so the default linuxspi entry is now automatically shipped if the option is enabled by configure. (We did similar already with the parport stuff in the past.)

So thanks to everyone.
